Segmentation fault with ifort

Last week I noticed that if I compile gimic with ifort current density calculations crash with a seg fault but bond integration works fine. So, I compiled with traceback to find out what's going on, since with gcc the calculations finish. 

```
./setup --fc=ifort --extra-fc-flags='-traceback -check all' --cc=icc --cxx=icc
```

I noticed two things. 

1) The ``rank`` variable in the file below is not initialized but used in a subroutine. Probably it needs to be defined as intent in. 

```
forrtl: severe (194): Run-Time Check Failure. The variable 'initialize$RANK$_114' is being used in '/homeappl/home/mariavd/appl_taito/gimic/src/fgimic/gimic.F90(96,13)' without being defined
Image              PC                Routine            Line        Source
gimic.bin          00000000004063E8  gimic_IP_initiali          96  gimic.F90
gimic.bin          000000000040460A  MAIN__                     44  gimic.F90
gimic.bin          000000000040455E  Unknown               Unknown  Unknown
libc-2.12.so       0000000005C4ED1D  __libc_start_main     Unknown  Unknown
gimic.bin          00000000004043E9  Unknown               Unknown  Unknown
```

2) In the jvector_plots subroutine the pointer to the tensor jtens is used but not allocated. The function get_acid(r, tens) takes the pointer to the tensor as intent(in), so it receives an undefined pointer. This problematic subroutine is called by the run_cdens subroutine. I actually don't know how the code can work when compiled with gcc with this issue. 

```
forrtl: severe (408): fort: (7): Attempt to use pointer JTENS when it is not associated with a target

Image              PC                Routine            Line        Source
gimic.bin          00000000004BDB60  Unknown               Unknown  Unknown
gimic.bin          000000000045466B  jfield_class_mp_j         326  jfield.f90
gimic.bin          0000000000405689  gimic_IP_run_cden         203  gimic.F90
gimic.bin          0000000000405B5E  gimic_IP_driver_          178  gimic.F90
gimic.bin          000000000040460F  MAIN__                     45  gimic.F90
gimic.bin          000000000040455E  Unknown               Unknown  Unknown
libc-2.12.so       00007FC3D9DC2D1D  __libc_start_main     Unknown  Unknown
gimic.bin          00000000004043E9  Unknown               Unknown  Unknown
```

Just for the sake of being thorough, compiling with gfortran and ``-Wall`` also points out the problem. You can replicate it by compiling with:
```
./setup  --extra-fc-flags='-g -O0 -Wall -Wextra -fbacktrace -fbounds-check -Wtabs -Wno-unused'
#
# Among the output:
#
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/parallel.F90:19.31:

    function init_mpi(is_mpirun) result(mpirank)
                               1
Warning: Dummy argument 'is_mpirun' at (1) was declared INTENT(OUT) but was not set
#
#
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:230.37:

    subroutine jvector_plots(this,mol,tag)
                                     1
Warning: Unused dummy argument 'mol' at (1)
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90: In function ‘jvector_plots’:
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:326:0: warning: ‘jtens.dim[0].lbound’ may be used uninitialized in this function [-Wmaybe-uninitialized]
                       val = get_acid(rr,jtens(:,idx))
 ^
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:326:0: warning: ‘jtens.dim[0].ubound’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:326:0: warning: ‘jtens.dim[0].stride’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:326:0: warning: ‘jtens.dim[1].lbound’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/homeappl/home/mariavd/gimic-upstream/gimic/src/fgimic/jfield.f90:326:0: warning: ‘jtens.dim[1].stride’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

There is a huge number of unused variables and function arguments that are not in use at all. Some day it would be good to have them fixed but for the time being they are not a priority. 

I am still puzzled how ACID works with an uninitialized pointer to the jvector in the case when the error is not caught by the compiler.

I will add the flags in the debug build which will be used when running the tests with valgrind.

--- a/src/fgimic/getkw.f90
+++ b/src/fgimic/getkw.f90
@@ -421,7 +421,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_INT) then
+            if (ptr%typ  /= KW_INT) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%ival
@@ -460,7 +460,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_IVEC) then
+            if (ptr%typ  /= KW_IVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%ivec
@@ -487,7 +487,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_DVEC) then
+            if (ptr%typ  /= KW_DVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%dvec
@@ -526,7 +526,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_LVEC) then
+            if (ptr%typ  /= KW_LVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%lvec
@@ -565,7 +565,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_STR) then
+            if (ptr%typ  /= KW_STR) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%str
@@ -612,7 +612,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_DBL) then
+            if (ptr%typ  /= KW_DBL) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%dval
@@ -639,7 +639,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_BOOL) then
+            if (ptr%typ  /= KW_BOOL) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%bool
@@ -668,7 +668,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_INT) then
+            if (ptr%typ  /= KW_INT) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%ival
@@ -696,7 +696,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_IVEC) then
+            if (ptr%typ  /= KW_IVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%ivec
@@ -723,7 +723,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_DVEC) then
+            if (ptr%typ  /= KW_DVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%dvec
@@ -750,7 +750,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_LVEC) then
+            if (ptr%typ  /= KW_LVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%lvec
@@ -777,7 +777,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_LVEC) then
+            if (ptr%typ  /= KW_LVEC) then
                 call getkw_typerr(ptr)
             else
                 val=>ptr%key%str
@@ -804,7 +804,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_DBL) then
+            if (ptr%typ  /= KW_DBL) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%dval
@@ -831,7 +831,7 @@ contains
 !                call getkw_errundef(ptr%id)
                 return
             end if
-            if (ptr%typ	 /= KW_BOOL) then
+            if (ptr%typ  /= KW_BOOL) then
                 call getkw_typerr(ptr)
             else
                 val=ptr%key%bool
@@ -1157,7 +1157,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_INT) then
+            if (ptr%typ  /= KW_INT) then
                 call getkw_typerr(ptr)
             else
                 ptr%key%ival=val
@@ -1181,7 +1181,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_IVEC) then
+            if (ptr%typ  /= KW_IVEC) then
                 call getkw_typerr(ptr)
             else
                 if (associated(ptr%key%ivec)) then
@@ -1209,7 +1209,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_DVEC) then
+            if (ptr%typ  /= KW_DVEC) then
                 call getkw_typerr(ptr)
             else
                 if (associated(ptr%key%dvec)) then
@@ -1237,7 +1237,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_DBL) then
+            if (ptr%typ  /= KW_DBL) then
                 call getkw_typerr(ptr)
             else
                 ptr%key%dval=val
@@ -1261,7 +1261,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_LVEC) then
+            if (ptr%typ  /= KW_LVEC) then
                 call getkw_typerr(ptr)
             else
                 if (associated(ptr%key%lvec)) then
@@ -1289,7 +1289,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_BOOL) then
+            if (ptr%typ  /= KW_BOOL) then
                 call getkw_typerr(ptr)
             else
                 ptr%key%bool=val
@@ -1313,7 +1313,7 @@ contains
 
         ok=findkw(this%active, path, ptr)
         if (ok) then
-            if (ptr%typ	 /= KW_LVEC) then
+            if (ptr%typ  /= KW_LVEC) then
                 call getkw_typerr(ptr)
             else
                 if (associated(ptr%key%str)) then

--- a/src/fgimic/gimic.F90
+++ b/src/fgimic/gimic.F90
@@ -196,17 +196,17 @@ contains
         if (settings%dryrun) return
 
         call calc_jvectors(jf, mol, xdens)
-        call jvector_plots(jf,mol,'')
+        call jvector_plots(jf,'')
 
         if (settings%is_uhf) then
             call calc_jvectors(jf, mol, xdens, 'alpha')
-            call jvector_plots(jf,mol, 'alpha')
+            call jvector_plots(jf, 'alpha')
 
             call calc_jvectors(jf, mol, xdens, 'beta')
-            call jvector_plots(jf,mol, 'beta')
+            call jvector_plots(jf, 'beta')
 
             call calc_jvectors(jf, mol, xdens, 'spindens')
-            call jvector_plots(jf,mol, 'spindens')
+            call jvector_plots(jf, 'spindens')
         endif
         call del_jfield(jf)
     end subroutine

--- a/src/fgimic/gimic.F90
+++ b/src/fgimic/gimic.F90
@@ -56,7 +56,7 @@ program gimic
 
 contains
     subroutine initialize()
-        integer(I4) :: i, hostnm, rank, ierr
+        integer(I4) :: i, hostnm, ierr
         integer(I4) :: chdir, system
         character(BUFLEN) :: title, fdate, sys
         real(DP), dimension(3) :: magnet
@@ -93,11 +93,11 @@ contains
 
         ierr=hostnm(sys)
         if (mpi_rank == 0) then
-            write(str_g, '(a,i3,a,a)') 'MPI master', rank,' on ', trim(sys)
+            write(str_g, '(a,i3,a,a)') 'MPI master', mpi_rank,' on ', trim(sys)
             call msg_debug(str_g,2)
         else
             call set_teletype_unit(DEVNULL)
-            write(str_g, '(a,i3,a,a)') 'MPI slave', rank,' on ', trim(sys)
+            write(str_g, '(a,i3,a,a)') 'MPI slave', mpi_rank,' on ', trim(sys)
             call msg_debug(str_g,2)
         end if
 

--- a/src/fgimic/grid.f90
+++ b/src/fgimic/grid.f90
@@ -426,7 +426,7 @@ contains
             call nl
             return
         end if
-99		format(a,3f12.8,a)
+99      format(a,3f12.8,a)
     end subroutine
 
     subroutine get_grid_size(this, i, j, k)

--- a/src/fgimic/grid.f90
+++ b/src/fgimic/grid.f90
@@ -354,7 +354,7 @@ contains
         this%gauss=.false.
         this%npts(1)=nint(this%l(1)/this%step(1))+1
         this%npts(2)=nint(this%l(2)/this%step(2))+1
-        if (this%l(3) == 0.d0 .or. this%step(3) == 0.d0) then
+        if (abs(this%l(3)) < tiny(0.0d0) .or. abs(this%step(3)) < tiny(0.d0)) then
             this%npts(3)=1
         else
             this%npts(3)=nint(this%l(3)/this%step(3))+1

--- a/src/fgimic/grid.f90
+++ b/src/fgimic/grid.f90
@@ -76,7 +76,7 @@ contains
                 call extgrid(this)
                 return
             case ('std','base')
-                call setup_std_grid(this,mol)
+                call setup_std_grid(this)
             case ('bond')
                 call setup_bond_grid(this,mol)
             case default
@@ -135,9 +135,8 @@ contains
         call nl
     end subroutine
 
-    subroutine setup_std_grid(this,mol)
+    subroutine setup_std_grid(this)
         type(grid_t) :: this
-        type(molecule_t) :: mol
         type(atom_t), pointer :: atom
 
         integer(I4) :: i

--- a/src/fgimic/integral.f90
+++ b/src/fgimic/integral.f90
@@ -468,7 +468,7 @@ contains
                     r=sqrt(sum((rr-center)**2))
                     call ctensor(jt, rr, tt, spin)
                     ! attention: output of get_acid is in au !
-                    val = get_acid(rr,tt)
+                    val = get_acid(tt)
                     if ( r > bound ) then
                         w=0.d0
                     else

--- a/src/fgimic/integral.f90
+++ b/src/fgimic/integral.f90
@@ -44,6 +44,7 @@ contains
 
     subroutine del_integral(this)
         type(integral_t) :: this
+        nullify(this%grid)
     end subroutine
 
     subroutine integrate_current(this, mol, xdens, spinn)

--- a/src/fgimic/jfield.f90
+++ b/src/fgimic/jfield.f90
@@ -322,7 +322,7 @@ contains
                     ! case ACID
                     if (settings%acid) then
                       idx = i+(j-1)*p1+(k-1)*p1*p2
-                      val = get_acid(rr,jtens(:,idx))
+                      val = get_acid(jtens(:,idx))
                     end if
                 end do
                     if (debug) then
@@ -548,7 +548,7 @@ contains
                     !v=buf(:,i+(j-1)*p1+(k-1)*p1*p2)
                     rr=gridpoint(this%grid,i,j,k)
                     idx = i+(j-1)*p1 + (k-1)*p1*p2
-                    val(i,j,k)= get_acid(rr, jtens(:,idx))
+                    val(i,j,k)= get_acid(jtens(:,idx))
                 end do
             end do
         end do

--- a/src/fgimic/jfield.f90
+++ b/src/fgimic/jfield.f90
@@ -227,11 +227,10 @@ contains
         return
     end function
 
-    subroutine jvector_plots(this,mol,tag)
+    subroutine jvector_plots(this,tag)
         use vtkplot_module
 
         type(jfield_t), intent(inout) :: this
-        type(molecule_t) :: mol
         character(*), optional :: tag
         logical :: circle_log
         logical :: debug 

--- a/src/fgimic/jfield.f90
+++ b/src/fgimic/jfield.f90
@@ -534,18 +534,13 @@ contains
         jtens => this%tens
         mag = this%b
 
-        !buf => this%vec
         maxi=0.d0
         mini=0.d0
         l=0
         idx = 0
-        !do i=1,p1
-        !    do j=1,p2
-        !        do k=1,p3
         do k=1,p3
             do j=1,p2
                 do i=1,p1
-                    !v=buf(:,i+(j-1)*p1+(k-1)*p1*p2)
                     rr=gridpoint(this%grid,i,j,k)
                     idx = i+(j-1)*p1 + (k-1)*p1*p2
                     val(i,j,k)= get_acid(jtens(:,idx))

--- a/src/fgimic/jfield.f90
+++ b/src/fgimic/jfield.f90
@@ -321,6 +321,7 @@ contains
                     end if
                     ! case ACID
                     if (settings%acid) then
+                      jtens => this%tens
                       idx = i+(j-1)*p1+(k-1)*p1*p2
                       val = get_acid(jtens(:,idx))
                     end if

--- a/src/fgimic/parallel.F90
+++ b/src/fgimic/parallel.F90
@@ -35,9 +35,9 @@ contains
         mpirank = 0
 #endif
         if (mpi_world_size == 1) then
-            settings%is_mpirun = .false.
+            is_mpirun = .false.
         else
-            settings%is_mpirun = .true.
+            is_mpirun = .true.
         end if
     end function
 

--- a/src/libgimic/acid.f90
+++ b/src/libgimic/acid.f90
@@ -8,7 +8,7 @@ module acid_module
      implicit none
 
 contains
-     function get_acid(r, tens) result(dT2)
+     function get_acid(tens) result(dT2)
      ! purpose: do acid summation of T tensor
      ! see: R. Herges and D. Gruenich,
      !      J. Chem. Phys. A, 105, 3214 (2001)
@@ -24,7 +24,6 @@ contains
      ! assume: T tensor is written on tens(9:number of points)
      !         first column of T eq. first entries of tens(1:3,npts)
      !
-     real(DP), dimension(3), intent(in) :: r
      real(DP), dimension(9), intent(in) :: tens
      real(DP) :: dT2, tmp
      real(DP) :: xxmyy, yymzz, zzmxx, xypyx, xzpzx, yzpzy

--- a/src/libgimic/basis.f90
+++ b/src/libgimic/basis.f90
@@ -82,7 +82,7 @@ contains
         end if
 
         call write_xyz(this, 'mol.xyz')
-77		format(a,i4)
+77              format(a,i4)
     end subroutine
 !
 ! Loop over ctr and set up screening thresholds
@@ -205,8 +205,8 @@ contains
             write(tun, 21, advance='no') ctr%cc(i),' (',ctr%ncc(i),') '
             call nl
         end do
-20		format(f20.10)
-21		format(f13.10,a,f13.10,a)
+20              format(f20.10)
+21              format(f13.10,a,f13.10,a)
     end subroutine
 
 !
@@ -291,10 +291,10 @@ contains
         call nl
 
 55      format(i3,a,i3,a,a1,a)
-66		format(a,i3,a2,2a1)
-77		format(a,f8.4)
-88		format(a,i4,a,a2,a)
-99		format(a)
+66              format(a,i3,a2,2a1)
+77              format(a,f8.4)
+88              format(a,i4,a,a2,a)
+99              format(a)
     end subroutine
 
 !
@@ -469,21 +469,21 @@ contains
     end function
 
     ! integrate s-funcs for debugging purpouses...
-!	subroutine s_int(Ctr)
-!		type(contraction_t), intent(in) :: Ctr
+!       subroutine s_int(Ctr)
+!               type(contraction_t), intent(in) :: Ctr
 !
-!		real(DP) :: qq
-!		integer(I4) :: i,j
+!               real(DP) :: qq
+!               integer(I4) :: i,j
 !
-!		qq=0.d0
-!		do i=1,Ctr%npf
-!			do j=1,Ctr%npf
-!				qq=qq+Ctr%ncc(i)*Ctr%ncc(j)*(PI/(Ctr%xp(i)+Ctr%xp(j)))**1.5
-!			end do
-!		end do
-!		print *, 's type integral:', qq
+!               qq=0.d0
+!               do i=1,Ctr%npf
+!                       do j=1,Ctr%npf
+!                               qq=qq+Ctr%ncc(i)*Ctr%ncc(j)*(PI/(Ctr%xp(i)+Ctr%xp(j)))**1.5
+!                       end do
+!               end do
+!               print *, 's type integral:', qq
 !
-!	end subroutine
+!       end subroutine
     subroutine write_xyz(mol, fname)
         type(molecule_t) :: mol
         character(*), intent(in) :: fname

--- a/src/libgimic/bfeval.f90
+++ b/src/libgimic/bfeval.f90
@@ -71,10 +71,10 @@ contains
         endif
         if (present(db)) then
             call mkdbop(this, r)
-            call dfdb(this, r, bf, db)
+            call dfdb(this, bf, db)
         endif
         if (present(d2)) then
-            call d2fdrdb(this, r,  bf, dr, d2)
+            call d2fdrdb(this, bf, dr, d2)
         endif
     end subroutine
 
@@ -188,9 +188,8 @@ contains
         end do
     end subroutine
 
-    subroutine d2fdrdb(this, r,  bfvec, drvec, thisv)
+    subroutine d2fdrdb(this, bfvec, drvec, thisv)
         type(bfeval_t) :: this
-        real(DP), dimension(:), intent(in) :: r
         real(DP), dimension(:), intent(in) :: bfvec
         real(DP), dimension(:,:), intent(in) :: drvec
         real(DP), dimension(:,:), pointer :: thisv
@@ -269,9 +268,8 @@ contains
         end do
     end subroutine
 
-    subroutine dfdb(this, r, bfvec, dbv)
+    subroutine dfdb(this, bfvec, dbv)
         type(bfeval_t) :: this
-        real(DP), dimension(:), intent(in) :: r
         real(DP), dimension(:), intent(in) :: bfvec
         real(DP), dimension(:,:), pointer :: dbv
 

--- a/src/libgimic/cao2sao.f90
+++ b/src/libgimic/cao2sao.f90
@@ -122,23 +122,8 @@ contains
     end subroutine
 
     ! Calculate the normalization factor for SGTO_lm
-    function nslm(l,m) result(q)
-        integer(I4), intent(in) :: l, m
-
-        real(DP) :: q, fac
-
-        q=1.d0/(2.d0**am)
-        q=q/dble(fact(l))
-        fac=2.d0*dble(fact(l+am)*fact(l-am))
-        if (am == 0) then
-            fac=fac/2.d0
-        end if
-        q=q*sqrt(fac)
-    end function
-
-    ! Calculate the normalization factor for SGTO_lm
-    function nslm2(l,m) result(q)
-        integer(I4), intent(in) :: l, m
+    function nslm(l) result(q)
+        integer(I4), intent(in) :: l
 
         real(DP) :: q, fac
 
@@ -152,8 +137,8 @@ contains
     end function
 
     ! Calculate the expansion coefficient for SGTO_lm
-    function clmtuv(l,m,t,u,v) result(q)
-        integer(I4), intent(in) :: l, m, t, u
+    function clmtuv(l,t,u,v) result(q)
+        integer(I4), intent(in) :: l, t, u
         real(DP) :: v
         integer(I4) :: iq
 
@@ -187,7 +172,7 @@ contains
             am=abs(m)
             vm=0.0
             if ( m < 0 ) vm=0.5d0
-            xnslm=nslm(l,m)
+            xnslm=nslm(l)
             do t=0,(l-am)/2
                 do u=0,t
                     do v=0,int(am*0.5d0-vm)
@@ -196,7 +181,7 @@ contains
                         j=int(2.d0*(dble(u)+vr))
                         k=l-2*t-am
                         n=gtomap(i,j,k)
-                        xop(idx,n)=clmtuv(l,m,t,u,vr)*xnslm
+                        xop(idx,n)=clmtuv(l,t,u,vr)*xnslm
                     end do
                 end do
             end do

--- a/src/libgimic/gimic_interface.f90
+++ b/src/libgimic/gimic_interface.f90
@@ -152,8 +152,13 @@ contains
 
     subroutine gimic_calc_modj(r, d) bind(c)
         real(C_DOUBLE), dimension(3), intent(in) :: r
+        real(8), dimension(3) :: r_
         real(C_DOUBLE), intent(out) :: d
         stop 'gimic_calc_modj(): NOT IMPLEMENTED YET!'
+
+        ! this is silly but just to avoid compiler warning that r is unused
+        r_ = r
+
         d = 0.0
     end subroutine
 

--- a/src/libgimic/globals.f90
+++ b/src/libgimic/globals.f90
@@ -160,7 +160,7 @@ contains
             read(fd,*,end=100)
             n=n+1
         end do
-100		rewind(fd)
+100     rewind(fd)
     end function
 
     function getfsize(fil) result(fs)

--- a/src/libgimic/intgrl.f90
+++ b/src/libgimic/intgrl.f90
@@ -78,7 +78,7 @@ contains
             allocate(b%ctr(bdim(i))) ! mem for all contrs. for atom i
             allocate(b%thrs(bdim(i))) ! mem for screening thrs. for atom i
             allocate(b%pos(bdim(i))) ! mem for pos idx thrs. for ctrs
-            call read_segs(b)    	! atom i
+            call read_segs(b) ! atom i
             b%nctr=bdim(i)
         end do
 

--- a/src/libgimic/jtensor.F90
+++ b/src/libgimic/jtensor.F90
@@ -186,6 +186,11 @@ contains
             end if
             dpd(b)=diapam*this%rho(b) ! diamag. contr. to J
             do m=1,3 !dm <x,y,z>
+
+              ! we zero these out to avoid compiler warning us that these may be used uninitialized
+              prsp1 = 0.0d0
+              prsp2 = 0.0d0
+
               if (settings%use_giao) then
 #ifdef HAVE_BLAS
                 ! (-i)**2 = -1

